format/udf-spec.md

```diff
@@ -0,0 +1,402 @@
+
+## Goals
+
+* Define a portable metadata format for both scalar and table SQL UDFs. The metadata is self-contained and can be moved across catalogs.
+* Support function evolution through versioning and rollback.
+* Provide consistent semantics for representing UDFs across engines.
+
+## Overview
+
+UDF metadata follows the same design principles as Iceberg table and view metadata: each function is represented by a
+**self-contained metadata file**. Metadata captures definitions, parameters, return types, documentation, security,
+properties, and engine-specific representations.
+
+* Any modification (new definition, updated representation, changed properties, etc.) creates a new metadata file, and atomically swaps in the new file as the current metadata.
+* Each metadata file includes recent definition versions, enabling rollbacks without external state.
+
+## Specification
+
+### UDF Metadata
+The UDF metadata file has the following fields:
+
+| Requirement | Field name        | Type                   | Description                                                                                                     |
+|-------------|-------------------|------------------------|-----------------------------------------------------------------------------------------------------------------|
+| *required*  | `function-uuid`   | `string`               | A UUID that identifies the function, generated once at creation.                                                |
+| *required*  | `format-version`  | `int`                  | Metadata format version (must be `1`).                                                                          |
+| *required*  | `definitions`     | `list<definition>`     | List of function [definition](#definition) entities.                                                            |
+| *required*  | `definition-log`  | `list<definition-log>` | History of [definition snapshots](#definition-log).                                                             |
+| *required*  | `parameter-names` | `list<parameter-name>` | Global ordered parameter names shared across all overloads. Overloads must use a prefix of this list, in order. |
+| *optional*  | `location`        | `string`               | Storage location of metadata files.                                                                             |
+| *optional*  | `properties`      | `map<string,string>`   | A string-to-string map of properties.                                                                           |
+| *optional*  | `secure`          | `boolean`              | Whether it is a secure function. Default: `false`.                                                              |
+| *optional*  | `doc`             | `string`               | Documentation string.                                                                                           |
+
+### Parameter-Name
```

Comment:
the notes for `secure` and `properties` should belong to the previous section. should the `Parameter-Name` subsection be moved after the `Notes:`? the notes for the `parameter-names` can be moved to that section too.

Reply:
Shouldn't this be `parameter` that has a `name` and a `doc`? A parameter name that contains a name is redundant.